import sys
import os
import bcrypt
# 1. Ø£Ø¶ÙÙ†Ø§ QDesktopWidget Ù‡Ù†Ø§
from PyQt5.QtWidgets import QWidget, QApplication, QVBoxLayout, QLabel, QGraphicsDropShadowEffect, QSizePolicy, QDesktopWidget
from PyQt5.uic import loadUi
from PyQt5.QtCore import pyqtSignal, Qt
from PyQt5.QtGui import QPixmap, QColor
from db import db

class Session:
    username = None
    role = None

class LoginWindow(QWidget):
    loginSuccess = pyqtSignal()

    def __init__(self):
        super().__init__()
        try:
            loadUi("login.ui", self)
        except Exception as e:
            print(f"Error loading UI: {e}")
            return
        
        # Ø¶Ø¨Ø· Ø­Ø¬Ù… Ø§Ù„Ù†Ø§ÙØ°Ø©
        self.resize(1000, 700)
        
        # -----------------------------------------------------
        # ğŸ”¥ Ø§Ù„Ø³Ø­Ø± Ù‡Ù†Ø§: ÙƒÙˆØ¯ ØªÙˆØ³ÙŠØ· Ø§Ù„Ù†Ø§ÙØ°Ø© ÙÙŠ Ù…Ù†ØªØµÙ Ø§Ù„Ø´Ø§Ø´Ø© ğŸ”¥
        # -----------------------------------------------------
        self.center_window()
        # -----------------------------------------------------

        # 1. Ø¶Ø¨Ø· Ù†Ø³Ø¨ Ø§Ù„ØªÙ‚Ø³ÙŠÙ… (70% ØµÙˆØ±Ø© - 30% Ù†Ù…ÙˆØ°Ø¬)
        if hasattr(self, "horizontalLayout"):
            self.horizontalLayout.setStretch(0, 7) 
            self.horizontalLayout.setStretch(1, 3)

        # 2. ÙˆØ¶Ø¹ Ø§Ù„ØµÙˆØ±Ø© ÙˆØ¥ØµÙ„Ø§Ø­ Ù…Ø´ÙƒÙ„Ø© Ø§Ù„Ø­Ø¬Ù… Ø§Ù„ÙƒØ¨ÙŠØ±
        current_dir = os.path.dirname(os.path.abspath(__file__))
        img_path = os.path.join(current_dir, "login_bg.jpeg") 
        
        if hasattr(self, "imagePanel"):
            layout = QVBoxLayout(self.imagePanel)
            layout.setContentsMargins(0, 0, 0, 0)
            
            bg_label = QLabel()
            bg_label.setScaledContents(True) 
            bg_label.setSizePolicy(QSizePolicy.Ignored, QSizePolicy.Ignored)
            
            if os.path.exists(img_path):
                bg_label.setPixmap(QPixmap(img_path))
            else:
                print(f"âŒ Image not found at: {img_path}")
                bg_label.setStyleSheet("background-color: #0F4D39; color: white;")
                bg_label.setText("Image Not Found")
                bg_label.setAlignment(Qt.AlignCenter)
            
            layout.addWidget(bg_label)

        # 3. Ø¥Ø¶Ø§ÙØ© Ø¸Ù„ Ù„Ù„Ø²Ø±
        if hasattr(self, "loginBtn"):
            shadow = QGraphicsDropShadowEffect()
            shadow.setBlurRadius(20)
            shadow.setXOffset(0)
            shadow.setYOffset(5)
            shadow.setColor(QColor(15, 77, 57, 80)) 
            self.loginBtn.setGraphicsEffect(shadow)

        # Ø±Ø¨Ø· Ø§Ù„Ø¥Ø´Ø§Ø±Ø§Øª
        self.loginBtn.clicked.connect(self.handle_login)
        self.passwordInput.returnPressed.connect(self.handle_login)

    def center_window(self):
        """Ø¯Ø§Ù„Ø© Ù„Ø­Ø³Ø§Ø¨ Ù…Ù†ØªØµÙ Ø§Ù„Ø´Ø§Ø´Ø© ÙˆÙˆØ¶Ø¹ Ø§Ù„Ù†Ø§ÙØ°Ø© ÙÙŠÙ‡Ø§"""
        # Ù†Ø­ØµÙ„ Ø¹Ù„Ù‰ Ù‡Ù†Ø¯Ø³Ø© Ø§Ù„Ù†Ø§ÙØ°Ø© Ø§Ù„Ø­Ø§Ù„ÙŠØ©
        qr = self.frameGeometry()
        # Ù†Ø­ØµÙ„ Ø¹Ù„Ù‰ Ù†Ù‚Ø·Ø© Ø§Ù„Ù…Ù†ØªØµÙ Ù„Ù„Ø´Ø§Ø´Ø©
        cp = QDesktopWidget().availableGeometry().center()
        # Ù†Ø­Ø±Ùƒ Ù…Ø±ÙƒØ² Ù…Ø³ØªØ·ÙŠÙ„ Ø§Ù„Ù†Ø§ÙØ°Ø© Ø¥Ù„Ù‰ Ù…Ù†ØªØµÙ Ø§Ù„Ø´Ø§Ø´Ø©
        qr.moveCenter(cp)
        # Ù†Ø­Ø±Ùƒ Ø§Ù„Ù†Ø§ÙØ°Ø© Ù†ÙØ³Ù‡Ø§ Ø¥Ù„Ù‰ Ø§Ù„Ø²Ø§ÙˆÙŠØ© Ø§Ù„Ø¹Ù„ÙŠØ§ Ø§Ù„ÙŠØ³Ø±Ù‰ Ù„Ù„Ù…Ø³ØªØ·ÙŠÙ„ Ø§Ù„Ø¬Ø¯ÙŠØ¯
        self.move(qr.topLeft())

    def handle_login(self):
        username = self.usernameInput.text().strip()
        password = self.passwordInput.text()

        if not username or not password:
            if hasattr(self, "errorLabel"):
                self.errorLabel.setText("Ø§Ù„Ø±Ø¬Ø§Ø¡ Ø¥Ø¯Ø®Ø§Ù„ Ø§Ù„Ø¨ÙŠØ§Ù†Ø§Øª")
                self.errorLabel.setStyleSheet("color: #D32F2F;")
            return

        try:
            user_data = db.fetch_one("SELECT password_hash, role FROM users WHERE username = ?", (username,))
        except Exception as e:
            if hasattr(self, "errorLabel"): self.errorLabel.setText("Ø®Ø·Ø£ ÙÙŠ Ø§Ù„Ø§ØªØµØ§Ù„")
            print(e)
            return

        if user_data:
            try:
                stored_hash = user_data['password_hash']
                role = user_data['role']
            except:
                stored_hash = user_data[0]
                role = user_data[1]

            if bcrypt.checkpw(password.encode('utf-8'), stored_hash.encode('utf-8')):
                Session.username = username
                Session.role = role
                self.loginSuccess.emit()
                self.close()
            else:
                if hasattr(self, "errorLabel"): self.errorLabel.setText("ÙƒÙ„Ù…Ø© Ø§Ù„Ù…Ø±ÙˆØ± ØºÙŠØ± ØµØ­ÙŠØ­Ø©")
                self.passwordInput.clear()
        else:
             if hasattr(self, "errorLabel"): self.errorLabel.setText("Ø§Ù„Ù…Ø³ØªØ®Ø¯Ù… ØºÙŠØ± Ù…ÙˆØ¬ÙˆØ¯")

if __name__ == "__main__":
    app = QApplication(sys.argv)
    win = LoginWindow()
    win.show()
    sys.exit(app.exec_())
